compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -612,213 +603,94 @@ impl<'hir> LoweringContext<'_, 'hir> {
-            constness: hir::Constness::NotConst,
-        }))
-    }
+        let coroutine_kind = hir::CoroutineKind::Desugared(desugaring_kind, coroutine_source);
+
+        // The `async` desugaring takes a resume argument and maintains a `task_context`,
+        // whereas a generator does not.
+        let (inputs, params, task_context): (&[_], &[_], _) = if match desugaring_kind {
+            hir::CoroutineDesugaring::Async => true,
+            hir::CoroutineDesugaring::Gen => false,
+            hir::CoroutineDesugaring::AsyncGen => true,
+        } {
+            // Resume argument type: `ResumeTy`
+            let unstable_span = self.mark_span_with_reason(
+                DesugaringKind::Async,
+                self.lower_span(span),
+                Some(self.allow_gen_future.clone()),
+            );
+            let resume_ty = self.make_lang_item_qpath(hir::LangItem::ResumeTy, unstable_span);
+            let input_ty = hir::Ty {
+                hir_id: self.next_id(),
+                kind: hir::TyKind::Path(resume_ty),
+                span: unstable_span,
+            };
+            let inputs = arena_vec![self; input_ty];
 
-    /// Lower a `async gen` construct to a generator that implements `AsyncIterator`.
-    ///
-    /// This results in:
-    ///
-    /// ```text
-    /// static move? |_task_context| -> () {
-    ///     <body>
-    /// }
-    /// ```
-    pub(super) fn make_async_gen_expr(
-        &mut self,
-        capture_clause: CaptureBy,
-        closure_node_id: NodeId,
-        _yield_ty: Option<hir::FnRetTy<'hir>>,
-        span: Span,
-        async_coroutine_source: hir::CoroutineSource,
-        body: impl FnOnce(&mut Self) -> hir::Expr<'hir>,
-    ) -> hir::ExprKind<'hir> {
-        let output = hir::FnRetTy::DefaultReturn(self.lower_span(span));
+            // Lower the argument pattern/ident. The ident is used again in the `.await` lowering.
+            let (pat, task_context_hid) = self.pat_ident_binding_mode(
+                span,
+                Ident::with_dummy_span(sym::_task_context),
+                hir::BindingAnnotation::MUT,
+            );
+            let param = hir::Param {
+                hir_id: self.next_id(),
+                pat,
+                ty_span: self.lower_span(span),
+                span: self.lower_span(span),
+            };
+            let params = arena_vec![self; param];
 
-        // Resume argument type: `ResumeTy`
-        let unstable_span = self.mark_span_with_reason(
-            DesugaringKind::Async,
-            self.lower_span(span),
-            Some(self.allow_gen_future.clone()),
-        );
-        let resume_ty = self.make_lang_item_qpath(hir::LangItem::ResumeTy, unstable_span);
-        let input_ty = hir::Ty {
-            hir_id: self.next_id(),
-            kind: hir::TyKind::Path(resume_ty),
-            span: unstable_span,
+            (inputs, params, Some(task_context_hid))
+        } else {
+            (&[], &[], None)
         };
```

Comment:
I'm so sorry for the back and forth, but... why did you write it like this? 😭 

When I asked for a match, I meant like this:
```
        let (inputs, params, task_context): (&[_], &[_], _) = match desugaring_kind {
            hir::CoroutineDesugaring::Async | hir::CoroutineDesugaring::AsyncGen => {
                // Resume argument type: `ResumeTy`
                let unstable_span = self.mark_span_with_reason(
                    DesugaringKind::Async,
                    self.lower_span(span),
                    Some(self.allow_gen_future.clone()),
                );
                let resume_ty = self.make_lang_item_qpath(hir::LangItem::ResumeTy, unstable_span);
                let input_ty = hir::Ty {
                    hir_id: self.next_id(),
                    kind: hir::TyKind::Path(resume_ty),
                    span: unstable_span,
                };
                let inputs = arena_vec![self; input_ty];

                // Lower the argument pattern/ident. The ident is used again in the `.await` lowering.
                let (pat, task_context_hid) = self.pat_ident_binding_mode(
                    span,
                    Ident::with_dummy_span(sym::_task_context),
                    hir::BindingAnnotation::MUT,
                );
                let param = hir::Param {
                    hir_id: self.next_id(),
                    pat,
                    ty_span: self.lower_span(span),
                    span: self.lower_span(span),
                };
                let params = arena_vec![self; param];

                (inputs, params, Some(task_context_hid))
            }
            hir::CoroutineDesugaring::Gen => (&[], &[], None),
        };
```

Reply:
🤦🏻‍♂️ indeed, "make it a match" should have been obvious :-D